authorization/invitation/service/invitation_service.go

```diff
@@ -151,6 +166,40 @@ func (s *invitationServiceImpl) Issue(ctx context.Context, issuingUserId uuid.UU
 					return errors.NewInternalError(ctx, error)
 				}
 			}
+
+			notifications = append(notifications, invitationNotification{
+				invitation: inv,
+				roles:      invitation.Roles,
+			})
+
+		}
+
+		inviter, err := s.Repositories().Identities().Load(ctx, issuingUserId)
+		if err != nil {
+			return err
+		}
+
+		// Use the notification service to send invitation e-mails to the invited users, in a separate thread
+		// Currently we only support two types of invitations;
+		//
+		// 1) Invite user to team, membership only, no organization
+		// 2) Invite user to space, roles only, no organization
+		//
+		if inviteToIdentity != nil {
+			identityResource, err := s.Repositories().ResourceRepository().Load(ctx, inviteToIdentity.IdentityResourceID.String)
+			if err != nil {
+				return err
+			}
+
+			if identityResource.ResourceType.Name == authorization.IdentityResourceTypeTeam {
+				err = s.processTeamInviteNotifications(ctx, inviteToIdentity, inviter.User.FullName, notifications, witURL)
```

Comment:
We can't actually call invitation processing here because we are still in a transaction. The notification service is called asynchronously in a new goroutine but the notification will be sent regardless of success or failure of transaction.
So, here is what may happen:

1. We start transaction to send two invitations
1.1. We create the first invitation in DB
1.2. A new goroutine is created to send the invitation
1.3. We created the second invitation in DB
1.4. Something fails and we return an error.
2. Transaction is rolled back and both Invitation 1 and 2 are not comited to the DB. The invitations do not exist.
3. Meanwhile the invitation 1 was sent to the user (it's in a separate goroutine fully independent from the transaction).
4. User get the invitation but it fails when he tries to accept it because it doesn't exist in DB.

I would suggest create all invitations first in a transaction then in a new loop process them.

Update: You actually call WIT to get Space in the invite notification processing methods. So it's even worse. You cal a remote service within a transaction in the same goroutine. One more reason to move the notification processing out of the transaction :)

Reply:
Good catch, I've moved this out of the transaction block.